src/chain/chain_information/build.rs

```diff
@@ -0,0 +1,979 @@
+    pub fn function_name(&self) -> &'static str {
+        match self {
+            RuntimeCall::AuraApiSlotDuration => "AuraApi_slot_duration",
+            RuntimeCall::AuraApiAuthorities => "AuraApi_authorities",
+            RuntimeCall::BabeApiCurrentEpoch => "BabeApi_current_epoch",
+            RuntimeCall::BabeApiNextEpoch => "BabeApi_next_epoch",
+            RuntimeCall::BabeApiConfiguration => "BabeApi_configuration",
+            RuntimeCall::GrandpaApiAuthorities => "GrandpaApi_grandpa_authorities",
+            RuntimeCall::GrandpaApiCurrentSetId => "GrandpaApi_current_set_id",
+        }
+    }
+
+    /// Returns the list of parameters to pass when making the call.
+    ///
+    /// The actual parameters are obtained by putting together all the returned buffers together.
+    pub fn parameter_vectored(
+        &'_ self,
+    ) -> impl Iterator<Item = impl AsRef<[u8]> + Clone + '_> + Clone + '_ {
+        iter::empty::<Vec<u8>>()
+    }
+
+    /// Returns the list of parameters to pass when making the call.
+    ///
+    /// This function is a convenience around [`RuntimeCall::parameter_vectored`].
+    pub fn parameter_vectored_vec(&self) -> Vec<u8> {
+        Vec::new()
+    }
+}
+
+impl fmt::Debug for RuntimeCall {
+    fn fmt(&self, f: &mut fmt::Formatter) -> fmt::Result {
+        fmt::Debug::fmt(&self.function_name(), f)
+    }
+}
+
+impl ChainInformationBuild {
+    /// Starts a new chain information build process.
+    ///
+    /// # Panic
+    ///
+    /// Panics if a [`ConfigFinalizedBlockHeader::NonGenesis`] is provided, and the header has
+    /// number 0.
+    ///
+    pub fn new(config: Config) -> Self {
+        // TODO: document
+        if let ConfigFinalizedBlockHeader::NonGenesis { header, .. } =
+            &config.finalized_block_header
+        {
+            assert_ne!(header.number, 0);
+        }
+
+        // TODO: also check versions?
+        let mut runtime_has_aura = false;
+        let mut runtime_has_babe = false;
+        let mut runtime_has_grandpa = false;
+        let aura_api_name = blake2_rfc::blake2b::blake2b(8, &[], b"AuraApi");
+        let babe_api_name = blake2_rfc::blake2b::blake2b(8, &[], b"BabeApi");
+        let grandpa_api_name = blake2_rfc::blake2b::blake2b(8, &[], b"GrandpaApi");
+        for api in config.runtime.runtime_version().decode().apis {
+            if api.name_hash == aura_api_name.as_bytes() {
+                runtime_has_aura = true;
+            } else if api.name_hash == babe_api_name.as_bytes() {
+                runtime_has_babe = true;
+            } else if api.name_hash == grandpa_api_name.as_bytes() {
+                runtime_has_grandpa = true;
+            }
+        }
+
+        let inner = ChainInformationBuildInner {
+            finalized_block_header: config.finalized_block_header,
+            call_in_progress: None,
+            virtual_machine: Some(config.runtime),
+            runtime_has_aura,
```

Comment:
q: can runtime have multiple options? e.g. aura & babe? if not, why it's not an enum?

Reply:
Yes it can, in which case a `MultipleConsensusAlgorithms` error is returned once we try to build the chain info.

We could also instantly return an error at initialization, but it makes the API more annoying to use, and this is such a niche error that I don't think it matters when we return the error.